[Bill Chee] Duke Increments

BillChee123 opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

Commit:
Merge branch 'gradle'

Commit:
level-1

Commit:
level-2

Commit:
level-3

Commit:
level-4

Commit:
A-TextUiTesting, level-5

A-TextUiTesting, level-5

Commit:
level-6

Commit:
level-7

level-7

Commit:
branch-Level-8

Commit:
Merge branch 'branch-Level-7'

Commit:
Resolving merge conflicts

Commit:
A-MoreOOP

Commit:
A-JUnit

Commit:
A-Packages

A-Packages

Commit:
A-MoreOOP parser implementation

Commit:
branch-Level-9

Commit:
ClearCommand to clear all tasks

Na-Nazhou reviewed (commented):
Personally I feel all the `*.class` files should not be committed. Similarly `list.txt` should be put inside .gitignore as well. 
Reference: https://stackoverflow.com/questions/8236785/should-class-files-be-put-under-version-control#targetText=3%20Answers&targetText=Yes%2C%20.class%20files%20produced%20from,omit%20from%20your%20source%20control%3F

Na-Nazhou commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,33 @@
+        return this.isExit;
+    }
+
+    public void setExit(boolean exit) {
```
I think this boolean parameter should be named as `isExit` according to the style guide.

Na-Nazhou commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class Ui {
+    private String line;
+    Scanner sc;
```
Is there any reason that `sc` is not a private attribute? Do you need to access it anywhere outside the `Ui` class?

BillChee123 replied:
Thank you for your comments! It helped me improved my code quality.

Na-Nazhou commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,54 @@
+     * printing the given input to the interface.
+     * @param args iteration of string inputs that each represent a line to be printed.
+     */
+    public void printStatement(String... args) {
```
Nice usage of varargs and stream! I did not think of this :D

Na-Nazhou commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+ *
+ */
+public class Deadline extends Task {
+    private String by;
```
I believe `by` can be removed from this class since it is not used any where. Correct me if I am wrong :)

Na-Nazhou commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    public String fileFormat() {
+        return String.format("D | %s | %s | %s %s", isDoneString(), getDescription(),
+                date.format(DateTimeFormatter.ofPattern("dd/MM/yyyy")),
```
I think `DateTimeFormatter.ofPattern("dd/MM/yyyy")` (and other DateTime formatters as well) can be abstracted out since this formatter is used in both `fileFormat()` and the constructor. Maybe you can have another class handling parsing/formatting DateTime and it would be also easier if you wish to parse/format DateTime in other classes (such as Event) in future.

Na-Nazhou commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "[done]" : "[x]"); //return tick or X symbols
```
This comment does not reflect the code correctly. Maybe you need to update it.

Na-Nazhou commented on src/main/java/command/DoneCommand.java:
```diff
@@ -0,0 +1,40 @@
+     * @throws DukeException If any error is encountered.
+     */
+    public DoneCommand(String input) throws DukeException {
+        this.index = Integer.parseInt(input);
```
`NumberFormatException` should be caught here as well I think (similar to what you did in `DeleteCommand`)?

Na-Nazhou commented on src/main/java/command/ListCommand.java:
```diff
@@ -0,0 +1,38 @@
+     * Command to List the existing tasks.
+     */
+    public ListCommand() {
+        this.setExit(false);
```
Instead if set `isExit` to be `false` in every `Command` subclass (except for `ByeCommand`), an alternative implementation is to  default `isExit` to be false in the parent `Command` class and then you only need to set it to be true in `ExitCommand` class.

Na-Nazhou commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * Initially populates the Task List with the hard drive data (Input data file).
```
There should be a blank comment line between the description and the list of tags, according to the style guide :P

Na-Nazhou commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,56 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                // !sc.hasNext("bye")
```
This comment can be removed I think?

Commit:
Level-10

branch level-10 initial changes

Before Iteration 3

Part 3 GUI

Styling GUI

Level-10

Level-10

Aulud commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,94 @@
+ * Command to add task into data.
+ *
+ */
+public class AddCommand extends Command {
```
I like how the the commands handling addition of a `Todo`, `Deadline` or `Event` has been abstracted into a more generic `AddCommand`. Nice!

Aulud commented on src/main/java/command/ByeCommand.java:
```diff
@@ -0,0 +1,22 @@
+public class ByeCommand extends Command {
+
+    public ByeCommand() {
+        this.setExit(true);
```
I think it would be cleaner if the `ByeCommand` class simply `@Overrides` the `isExit` method (which by default would return `false`) from `Command` to return `true`, instead of maintaining and using an additional method `setExit` to set the parent's attribute.

BillChee123 replied:
Yes I looked through your code and agree that it isn't necessary to really use a isExit variable since it is already unique to each command class hence just need to implement each isExit() method differently. Thank you for your feedback and idea!

Aulud commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,33 @@
+     * @param ui duke Ui object responsible for interface response.
+     * @param storage duke Storage object to make changes to hard drive data.
+     */
+    public abstract void execute(TaskList tasklist, Ui ui, Storage storage) throws DukeException;
```
It may be worth considering changing the `execute` method of `Command` and its child classes such that executing the command returns a result of some type, which the main `Duke` class would pass to its `Ui` instance to handle. This would de-couple commands from the class that handles UI and allow it to be more flexible.

Aulud commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+    public DeleteCommand(String input) throws DukeException {
+        try {
+            this.index = Integer.parseInt(input);
+            this.setExit(false);
```
Calling `setExit(false)` in all command constructors is a lot of work - consider having the field set in the constructor of the abstract class (which will be invoked by the child classes) instead.

Aulud commented on src/main/java/command/FindCommand.java:
```diff
@@ -0,0 +1,32 @@
+    public void execute(TaskList tasklist, Ui ui, Storage storage) throws DukeException {
+        Stream<String> taskStream = tasklist
+                .getList()
+                .stream()
```
Nice stream solution! :)

Aulud commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,68 @@
+        String cmd = arr[0];
+        Command command;
+        switch (cmd) {
+        case "todo":
```
According to the Java Coding Standard,
> The explicit `//Fallthrough` comment should be included whenever there is a case statement without a break statement.

This affects line 31 - 32.

BillChee123 replied:
These cases actually have break statement. This is a syntax indicating that they will all execute the code after the event case.

Aulud commented on src/main/java/command/ClearCommand.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public void execute(TaskList tasklist, Ui ui, Storage storage) throws DukeException {
+        tasklist.clearAll();
```
Nice clear command!

Aulud commented on src/test/java/DeadlineTest.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Test
+    public void deadlineTest() {
+        Task deadline = new Deadline("Deadline deadlinetest", "23/05/1997", "1800");
```
Nice! Your tests are pretty thorough.

Aulud reviewed (commented):
Nice work overall!

Commit:
apply fxml changes in Level-10

FXML

FXML

FXML

Commit:
Merge branch 'fxml'

Commit:
set up Travis CI

Commit:
Add A-CodeQuality increment

Commit:
Add A-Assertions increment

Commit:
Add D-Trivia extension

Add D-Trivia extension

BillChee123 committed:
Merge pull request #4 from LivingOnEdge/A-CodeQuality-Increment

A code quality increment

BillChee123 committed:
Merge pull request #5 from LivingOnEdge/A-Assertions-Increment

A assertions increment

Commit:
Add D-Trivia VocabularyList refactoring

Add duke.txt and vocabulary.txt to .gitignore

Commit:
Add a brief User Guide

Update README.md

Update README.md

Commit:
Fix cross platform issues with jar file
